adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java

```diff
@@ -195,80 +192,354 @@ public void run() {
+     *
+     * @see <a href-"https://tools.ietf.org/html/rfc7617">RFC-7617</a>
+     * @see <a href="https://tools.ietf.org/html/rfc6750">RFC-6750</a>
+     */
+    private static class Challenge {
+
+        /**
+         * Regex sequence intended to be prefixed with another value. Whichever value precedes it
+         * will be parsed/grouped-out, assuming it is not bounded by a par of double-quotes ("").
+         */
+        private static final String REGEX_UNQUOTED_LOOKAHEAD = "(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)";
+
+        /**
+         * Regex sequence to parse unquoted equals (=) signs.
+         */
+        private static final String REGEX_SPLIT_UNQUOTED_EQUALS = "=" + REGEX_UNQUOTED_LOOKAHEAD;
+
+        /**
+         * Regex sequence to parse unquoted commas (,).
+         */
+        private static final String REGEX_SPLIT_UNQUOTED_COMMA = "," + REGEX_UNQUOTED_LOOKAHEAD;
+
+        /**
+         * Regex sequence to parse schemes from WWW-Authenticate header values.
+         */
+        private static final String REGEX_STRING_TOKEN_WITH_SCHEME = "^([^\\s|^=]+)[\\s|\\t]+([^=]*=[^=]*)+$";
+
+        /**
+         * Comma+space suffix used to preserve formatting during parsing.
+         */
+        private static final String SUFFIX_COMMA = ", ";
+
+        /**
+         * The authentication scheme of this challenge (ex. Basic, Bearer).
+         */
+        private String mScheme;
+
+        /**
+         * The parameters of this scheme (ex. realm, authoritization_uri, scope).
+         */
+        private Map<String, String> mParameters;
+
+        /**
+         * Constructs a new Challenge.
+         *
+         * @param scheme The scheme used by this challenge.
+         * @param params The params which accompany this challenge.
+         */
+        private Challenge(final String scheme, final Map<String, String> params) {
+            mScheme = scheme;
+            mParameters = params;
+        }
+
+        /**
+         * Gets the scheme.
+         *
+         * @return The scheme to get.
+         */
+        public String getScheme() {
+            return mScheme;
+        }
+
+        /**
+         * Gets the parameters.
+         *
+         * @return The parameters to get.
+         */
+        public Map<String, String> getParameters() {
+            return mParameters;
+        }
+
+        /**
+         * Parses a single challenge String, typically the value field of a WWW-Authenticate header.
+         *
+         * @param challenge The challenge String to parse.
+         * @return The Challenge object derived from challenge.
+         * @throws ResourceAuthenticationChallengeException If a parsing error is encountered or
+         *                                                  the String is malformed.
+         */
+        static Challenge parseChallenge(final String challenge) throws ResourceAuthenticationChallengeException {
+            final String scheme = parseScheme(challenge);
```

Comment:
do we need null checks for scheme, challengeSansScheme and challenge?

Reply:
Resolved in 78b52908d59b9e8fbac42ce896314e58063224da